File: easy/merge_sorted_array_test.go
package easy

import (
	"reflect"
	"testing"
)

func TestMergeSortedArray(t *testing.T) {
	cases := []struct {
		first      []int
		firstSize  int
		second     []int
		secondSize int
		expected   []int
	}{
		{[]int{1, 2, 3, 0, 0, 0}, 3, []int{2, 5, 6}, 3, []int{1, 2, 2, 3, 5, 6}},
		{[]int{1}, 1, []int{}, 0, []int{1}},
	}
	for _, c := range cases {
		t.Run("", func(t *testing.T) {
			result := MergeSortedArray(c.first, c.firstSize, c.second, c.secondSize)
			if !reflect.DeepEqual(result, c.expected) {
				t.Errorf("MergeSortedArray(%v, %d, %v, %d) == %v, expected %v", c.first, c.firstSize, c.second, c.secondSize, result, c.expected)
			}

		})

	}
}

func TestMergeSortedArrayInPlace(t *testing.T) {
	cases := []struct {
		first      []int
		firstSize  int
		second     []int
		secondSize int
		expected   []int
	}{
		{[]int{1, 2, 3, 0, 0, 0}, 3, []int{2, 5, 6}, 3, []int{1, 2, 2, 3, 5, 6}},
		{[]int{1}, 1, []int{}, 0, []int{1}},
	}
	for _, c := range cases {
		cache := c.first
		t.Run("", func(t *testing.T) {
			MergeSortedArrayInPlace(c.first, c.firstSize, c.second, c.secondSize)
			if !reflect.DeepEqual(c.first, c.expected) {
				t.Errorf("MergeSortedArrayInPlace(%v, %d, %v, %d), received: %v expected %v", cache, c.firstSize, c.second, c.secondSize, c.first, c.expected)
			}

		})

	}
}

File: src/test/java/easy/MergeSortedArrayTest.java
package easy;

import static org.junit.jupiter.api.Assertions.*;

import java.util.stream.Stream;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

public class MergeSortedArrayTest {
  MergeSortedArray instance = new MergeSortedArray();

  @ParameterizedTest
  @MethodSource("dataMergeSortedArray")
  public void testMergeSortedArray(
      int[] first, int firstSize, int[] second, int secondSize, int[] expected) {

    int[] result = instance.solve(first, firstSize, second, secondSize);

    assertArrayEquals(expected, result);
  }

  @ParameterizedTest
  @MethodSource("dataMergeSortedArray")
  public void testMergeSortedArrayInPlace( int[] first, int firstSize, int[] second, int secondSize, int[] expected) {
    instance.mergeSortedArrayInPlace(first, firstSize, second, secondSize);
    assertArrayEquals(expected, first);
  }

  public static Stream<Arguments> dataMergeSortedArray() {
    return Stream.of(
        Arguments.of(
            new int[] {1, 2, 3, 0, 0, 0}, 3, new int[] {2, 5, 6}, 3, new int[] {1, 2, 2, 3, 5, 6}),
        Arguments.of(new int[]{0}, 0, new int[]{1}, 1, new int[]{1}),
        Arguments.of(new int[] {1}, 1, new int[] {}, 0, new int[] {1}));
  }
}
